# MATH 544. MIDTERM

# Problem 1.

Let A be the matrix defined below.
* (a) Compute a lower triangular matrix L and upper triangular matrix U such that A = LU.
* (b) Make use of L and U to solve the system Ax = b.
* (c) Consider the iterative Jacobi method to solve the same system. Does it work? If not,think of preconditioning the system so that the Jacobi method works and perform 4 steps of the iteration. (Hint: Consider a permutation of the rows of A.)


Ax = b where

     [ 2 -1  1][x1]   [ 2]

     [ 6  1  2][x2] = [ 2]

     [ 2 -1  3][x3]   [ 0]


# Solution:

In [ ]:
# To solve Ax=b, We start by constructing the arrays for A and b.
import numpy as np
from numpy.linalg import eig

A = np.array([[ 2, -1,  1],
              [ 6,  1,  2],
              [ 2, -1,  3]])
b = np.transpose(np.array([ 2,  2,  0]))
# To solve the system we do

x = np.linalg.solve(A,b)
print("Real solution: ")
print(x)

Real solution: 
[ 0.875 -1.25  -1.   ]


**Using Jacobi Iteration:**

In [ ]:
A = np.array([[ 2, -1,  1],
              [ 6,  1,  2],
              [ 2, -1,  3]])

L = np.array([[ 0,  0,  0],
              [ 6,  0,  0],
              [ 2, -1,  0]])

D = np.array([[ 2,  0,  0],
              [ 0,  1,  0],
              [ 0,  0,  3]])

U = np.array([[ 0, -1,  1],
              [ 0,  0,  2],
              [ 0,  0,  0]])

D_inverse =  np. linalg. inv(D)
b = np.transpose(np.array([ 2,  2,  0]))
D_inverse_b = np.dot(D_inverse, b)

BJ = np.dot(-D_inverse, L+U)
w,v=eig(BJ)

print('BJ evalues:', w)

x0 = np.transpose(np.array([ 0,  0,  0]))
x = list();
x.append(x0)

for i in range(10):
  x1 = np.dot(BJ, x0)+ D_inverse_b
  x0 = x1
  x.append(x1)
Aproximations = np.array(x)
print(Aproximations)

BJ evalues: [-0.23452149+1.87038347j -0.23452149-1.87038347j  0.46904298+0.j        ]
[[   0.            0.            0.        ]
 [   1.            2.            0.        ]
 [   2.           -4.            0.        ]
 [  -1.          -10.           -2.66666667]
 [  -2.66666667   13.33333333   -2.66666667]
 [   9.           23.33333333    6.22222222]
 [   9.55555556  -64.44444444    1.77777778]
 [ -32.11111111  -58.88888889  -27.85185185]
 [ -14.51851852  250.37037037    1.77777778]
 [ 125.2962963    85.55555556   93.13580247]
 [  -2.79012346 -936.04938272  -55.01234568]]


The Jacobi Iteration applied to the given matrix A did not converge, as indicated by the solution vector [−2.7901,−936.0494,−55.0123], which shows magnitudes that are blowing up. This non-convergence occurs because the matrix
A is not well-preconditioned; specifically, partial pivoting has not been applied, meaning the largest absolute value in each column is not in the pivotal position.

To improve convergence, we will use partial pivoting by swapping rows 1 and 2 of A. This step will bring the largest absolute value in the first column to the top-left position, resulting in a preconditioned matrix that is better suited for iterative methods like the Jacobi Iteration.

After performing this row swap, we will proceed with the Jacobi Iteration on the new matrix, expecting it to yield a convergent solution due to the improved conditioning.

In [ ]:
A = np.array([[ 6,  1,  2],
              [ 2, -1,  1],
              [ 2, -1,  3]])

# P = np.array([[ 0,  1,  0],
#               [ 1,  0,  0],
#               [ 0,  0,  1]])

# A = np.dot(P, A)

L = np.array([[ 0,  0,  0],
              [ 2,  0,  0],
              [ 2, -1,  0]])

D = np.array([[ 6,  0,  0],
              [ 0, -1,  0],
              [ 0,  0,  3]])

U = np.array([[ 0,  1,  2],
              [ 0,  0,  1],
              [ 0,  0,  0]])

D_inverse =  np. linalg. inv(D)
b = np.transpose(np.array([ 2,  2,  0]))
# b = np.dot(P, b)
D_inverse_b = np.dot(D_inverse, b)

BJ = np.dot(-D_inverse, L+U)
w,v=eig(BJ)

print('BJ evalues:', w)

x0 = np.transpose(np.array([ 0,  0,  0]))
x = list();
x.append(x0)

for i in range(10):
  x1 = np.dot(BJ, x0)+ D_inverse_b
  x0 = x1
  x.append(x1)
Aproximations = np.array(x)
print(Aproximations)

BJ evalues: [-0.6310964+0.j          0.3155482+0.27656785j  0.3155482-0.27656785j]
[[ 0.          0.          0.        ]
 [ 0.33333333 -2.          0.        ]
 [ 0.66666667 -1.33333333 -0.88888889]
 [ 0.85185185 -1.55555556 -0.88888889]
 [ 0.88888889 -1.18518519 -1.08641975]
 [ 0.89300412 -1.30864198 -0.98765432]
 [ 0.88065844 -1.20164609 -1.03155007]
 [ 0.8774577  -1.2702332  -0.98765432]
 [ 0.87425697 -1.23273891 -1.00838287]
 [ 0.87491744 -1.25986892 -0.99375095]
 [ 0.8745618  -1.24391607 -1.0032346 ]]


As expected, the approximated soltuion [ 0.8745618  -1.24391607 -1.0032346 ] is approaching to the real solution [ 0.875 -1.25  -1.   ] that we've found earlier.

In [1]:
# !sudo apt-get install texlive-xetex texlive-fonts-recommended texlive-plain-generic

In [2]:
# !jupyter nbconvert --to pdf /content/Math544_Midterm_Lazizbek.ipynb